src/Development/IDE/Core/Shake.hs

```diff
@@ -268,30 +293,44 @@ instance NFData v => NFData (Value v)
 currentValue :: Value v -> Maybe v
 currentValue (Succeeded _ v) = Just v
 currentValue (Stale _ _) = Nothing
-currentValue Failed = Nothing
+currentValue Failed{} = Nothing
 
 -- | Return the most recent, potentially stale, value and a PositionMapping
 -- for the version of that value.
-lastValueIO :: ShakeExtras -> NormalizedFilePath -> Value v -> IO (Maybe (v, PositionMapping))
-lastValueIO ShakeExtras{positionMapping} file v = do
-    allMappings <- liftIO $ readVar positionMapping
-    pure $ case v of
-        Succeeded ver v -> Just (v, mappingForVersion allMappings file ver)
-        Stale ver v -> Just (v, mappingForVersion allMappings file ver)
-        Failed -> Nothing
+lastValueIO :: IdeRule k v => ShakeExtras -> k -> NormalizedFilePath -> IO (Maybe (v, PositionMapping))
+lastValueIO s@ShakeExtras{positionMapping,persistentKeys,state} k file = do
+  modifyVar state $ \hm -> do
+    let readPersistent = do
+          pmap <- readVar persistentKeys
+          mv <- runMaybeT $ do
+            liftIO $ Logger.logInfo (logger s) $ T.pack $ "LOOKUP UP PERSISTENT FOR" ++ show k
+            f <- MaybeT $ pure $ HMap.lookup (Key k) pmap
+            dv <- MaybeT $ runIdeAction "lastValueIO" s $ f file
```

Comment:
Isn't this a potential deadlock if `f file` tries to access the state?

Reply:
Yes, but not very different from a cycle in shake dependencies?

Reply:
But there is no cycle involved here, right? 

Put differently, why does this lookup+update need to be atomic?

Reply:
Hmm, I see

Reply:
Fixed I think. This bit requires some careful consideration.